<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Genome-wide pairwise inference of synonymous versus non-synonymous substitution rates (dnds) across multiple subject organisms — dnds_across_multiple_species • homologr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Genome-wide pairwise inference of synonymous versus non-synonymous substitution rates (dnds) across multiple subject organisms — dnds_across_multiple_species" />
<meta property="og:description" content="This function allows you to compute dnds maps between a query organism
and a set subject organisms stored in the same folder. The corresponding dnds maps are then stored in an output folder." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">homologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genome-wide pairwise inference of synonymous versus non-synonymous substitution rates (dnds) across multiple subject organisms</h1>
    
    <div class="hidden name"><code>dnds_across_multiple_species.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function allows you to compute dnds maps between a query organism
and a set subject organisms stored in the same folder. The corresponding dnds maps are then stored in an output folder.</p>
    </div>

    <pre class="usage"><span class='fu'>dnds_across_multiple_species</span>(
  <span class='kw'>query</span>,
  <span class='kw'>subjects_folder</span>,
  <span class='kw'>output_folder</span>,
  format = <span class='st'>"fasta"</span>,
  ortho_detection = <span class='st'>"DIAMOND-RBH"</span>,
  delete_corrupt_cds = <span class='fl'>FALSE</span>,
  store_locally = <span class='fl'>FALSE</span>,
  sensitivity_mode = <span class='st'>"ultra-sensitive"</span>,
  out_format = <span class='st'>"csv"</span>,
  evalue = <span class='st'>"1E-5"</span>,
  max_target_seqs = <span class='fl'>5000</span>,
  cores = <span class='fl'>1</span>,
  hard_mask = <span class='fl'>TRUE</span>,
  diamond_exec_path = <span class='kw'>NULL</span>,
  add_makedb_options = <span class='kw'>NULL</span>,
  add_diamond_options = <span class='kw'>NULL</span>,
  min_qry_coverage_hsp = <span class='fl'>50</span>,
  min_qry_perc_identity = <span class='fl'>10</span>,
  aa_aln_type = <span class='st'>"pairwise"</span>,
  aa_aln_tool = <span class='st'>"NW"</span>,
  codon_aln_tool = <span class='st'>"pal2nal"</span>,
  dnds_estimation = <span class='st'>"Li"</span>,
  progress_bar = <span class='fl'>TRUE</span>,
  diamond_hit_tables = <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span>(<span class='kw'>output_folder</span>, <span class='st'>"diamond_hit_tables"</span>),
  sep = <span class='st'>";"</span>,
  <span class='kw'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query</th>
      <td><p>a character string specifying the path to the CDS file of the query organism.</p></td>
    </tr>
    <tr>
      <th>subjects_folder</th>
      <td><p>a character string specifying the path to the folder where CDS files of the subject organisms are stored.</p></td>
    </tr>
    <tr>
      <th>output_folder</th>
      <td><p>a character string specifying the path to the folder where output dnds maps should be stored.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>a character string specifying the file format of the sequence file, e.g. <code>format = "fasta"</code>, <code>format = "gbk"</code>. See
<code><a href='import_proteome.html'>import_proteome</a></code> for more details.</p></td>
    </tr>
    <tr>
      <th>ortho_detection</th>
      <td><p>a character string specifying the orthology inference method that shall be performed to detect orthologous genes. Options are:</p><ul>
<li><p><code>ortho_detection ="DIAMOND-BH"</code>: DIAMOND unidirectional best hit.</p></li>
<li><p><code>ortho_detection = "DIAMOND-RBH"</code>: DIAMOND reciprocal/bidirectional best hit (default).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>delete_corrupt_cds</th>
      <td><p>a logical value indicating whether sequences with corrupt base triplets should be removed from the input <code>file</code>. This is the case when the length of coding sequences cannot be divided by 3 and thus the coding sequence contains at least one corrupt base triplet.</p></td>
    </tr>
    <tr>
      <th>store_locally</th>
      <td><p>a logical value indicating whether or not alignment files shall be stored locally rather than in <code><a href='https://rdrr.io/r/base/tempfile.html'>tempdir()</a></code>.</p></td>
    </tr>
    <tr>
      <th>sensitivity_mode</th>
      <td><p>specify the level of alignment sensitivity. The higher the sensitivity level, the more deep homologs can be found, but at the cost of reduced computational speed.</p><ul>
<li><p><code>sensitivity_mode = "fast"</code> : fastest alignment mode, but least sensitive (default). Designed for finding hits of &gt;70<!-- % identity and short read alignment such as NGS reads. --></p></li>
<li><p><code>sensitivity_mode = "mid-sensitive"</code> : fast alignments between the <code>fast</code> mode and the sensitive mode in sensitivity.</p></li>
<li><p><code>sensitivity_mode = "sensitive"</code> : fast alignments, but full sensitivity for hits &gt;40<!-- % identity. --></p></li>
<li><p><code>sensitivity_mode = "more-sensitive"</code> : more sensitive than the <code>sensitive</code> mode.</p></li>
<li><p><code>sensitivity_mode = "very-sensitive"</code> : sensitive alignment mode.</p></li>
<li><p><code>sensitivity_mode = "ultra-sensitive"</code> : most sensitive alignment mode (sensitivity as high as BLASTP).</p></li>
</ul></td>
    </tr>
    <tr>
      <th>out_format</th>
      <td><p>a character string specifying the format of the file in which the DIAMOND results shall be stored.
Available options are:</p><ul>
<li><p><code>out_format = "pair"</code> : Pairwise</p></li>
<li><p><code>out_format = "xml"</code> : XML</p></li>
<li><p><code>out_format = "csv"</code> : Comma-separated file</p></li>
</ul></td>
    </tr>
    <tr>
      <th>evalue</th>
      <td><p>a character string specifying the e-value for DIAMOND based orthology inference that is performed
in the process of dnds computations. Please use the scientific notation.</p></td>
    </tr>
    <tr>
      <th>max_target_seqs</th>
      <td><p>maximum number of aligned sequences that shall be retained. Please be aware that <code>max_target_seqs</code> selects best hits based on the database entry and not by the best e-value. See details here: https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty833/5106166 .</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>number of computing cores that shall be used to perform parallelized computations.</p></td>
    </tr>
    <tr>
      <th>hard_mask</th>
      <td><p>shall low complexity regions be hard masked with TANTAN? Default is <code>db_hard_mask = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>diamond_exec_path</th>
      <td><p>a path to the DIAMOND executable or <code>conda/miniconda</code> folder.</p></td>
    </tr>
    <tr>
      <th>add_makedb_options</th>
      <td><p>a character string specifying additional makedb options that shall be passed on to the diamond makedb command line call, e.g. <code>add_make_options = "--taxonnames"</code> (Default is <code>add_diamond_options = NULL</code>).</p></td>
    </tr>
    <tr>
      <th>add_diamond_options</th>
      <td><p>a character string specifying additional diamond options that shall be passed on to the diamond command line call, e.g. <code>add_diamond_options = "--block-size 4.0 --compress 1 --no-self-hits"</code> (Default is <code>add_diamond_options = NULL</code>).</p></td>
    </tr>
    <tr>
      <th>min_qry_coverage_hsp</th>
      <td><p>minimum <code>qcovhsp</code> (= query coverage of the HSP) of an orthologous hit (a value between 1 and 100).</p></td>
    </tr>
    <tr>
      <th>min_qry_perc_identity</th>
      <td><p>minimum <code>perc_identity</code> (= percent sequence identity between query and selected HSP) of an orthologous hit (a value between 1 and 100).</p></td>
    </tr>
    <tr>
      <th>aa_aln_type</th>
      <td><p>a character string specifying the amino acid alignement type: <code>aa_aln_type = "multiple"</code> or <code>aa_aln_type = "pairwise"</code>.
Default is <code>aa_aln_type = "pairwise"</code>.</p></td>
    </tr>
    <tr>
      <th>aa_aln_tool</th>
      <td><p>a character string specifying the program that should be used e.g. "clustalw".</p></td>
    </tr>
    <tr>
      <th>codon_aln_tool</th>
      <td><p>a character string specifying the codon alignment tool that shall be used.
Default is <code>codon_aln_tool = "pal2nal"</code>. Right now only "pal2nal" can be selected as codon alignment tool.</p></td>
    </tr>
    <tr>
      <th>dnds_estimation</th>
      <td><p>a character string specifying the dnds estimation method, e.g. <code>dnds_estimation = "Li"</code> (default).</p></td>
    </tr>
    <tr>
      <th>progress_bar</th>
      <td><p>should a progress bar be shown. Default is <code>progress_bar = TRUE</code>.</p></td>
    </tr>
    <tr>
      <th>diamond_hit_tables</th>
      <td><p>a file path to a folder where DIAMOND hit tables internally generated with <code><a href='diamond_best_hits.html'>diamond_best_hits</a></code> or <code><a href='diamond_reciprocal_best_hits.html'>diamond_reciprocal_best_hits</a></code> will be stored.</p></td>
    </tr>
    <tr>
      <th>sep</th>
      <td><p>a file separator that is used to store maps as csv file.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>additional parameters that shall be passed to  <code><a href='dnds.html'>dnds</a></code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Given a query organism and a set of subject organisms that are stored in the same folder,
this function crawls through all subject organism and as a first step computes the pairwise
dnds Maps between query and subject organism and as a second step stores the corresponding Map
in an output folder.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='dnds.html'>dnds</a></code>, <code><a href='import_dnds_across_multiple_species.html'>import_dnds_across_multiple_species</a></code>, <code><a href='import_dnds_tbl.html'>import_dnds_tbl</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hajk-Georg Drost</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'># running dnds across several species</span>
<span class='fu'>dnds_across_multiple_species</span>(
   query      = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, package = <span class='st'>'homologr'</span>),
   subjects_folder = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/map_gen_example'</span>, package = <span class='st'>'homologr'</span>),
   aa_aln_type      = <span class='st'>"pairwise"</span>,
   aa_aln_tool      = <span class='st'>"NW"</span>,
   codon_aln_tool   = <span class='st'>"pal2nal"</span>,
   dnds_estimation  = <span class='st'>"Li"</span>,
   output_folder   = <span class='st'>"homologr_dnds_maps"</span>,
   quiet           = <span class='fl'>TRUE</span>,
   cores      = <span class='fl'>1</span>
)

<span class='co'># running dnds across several species using DIAMOND executable</span>
<span class='co'># path '/opt/miniconda3/bin/'</span>
<span class='fu'>dnds_across_multiple_species</span>(
   query      = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, package = <span class='st'>'homologr'</span>),
   subjects_folder = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/map_gen_example'</span>, package = <span class='st'>'homologr'</span>),
   diamond_exec_path = <span class='st'>"/opt/miniconda3/bin/"</span>,
   aa_aln_type      = <span class='st'>"pairwise"</span>,
   aa_aln_tool      = <span class='st'>"NW"</span>,
   codon_aln_tool   = <span class='st'>"pal2nal"</span>,
   dnds_estimation  = <span class='st'>"Li"</span>,
   output_folder   = <span class='st'>"homologr_dnds_maps"</span>,
   quiet           = <span class='fl'>TRUE</span>,
   cores      = <span class='fl'>1</span>
)
}
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


