<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Import dnds tables by gene locus and splice varaint for a set of species — import_dnds_across_multiple_species • homologr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Import dnds tables by gene locus and splice varaint for a set of species — import_dnds_across_multiple_species" />
<meta property="og:description" content="Given an input folder storing dnds tables generated by dnds_across_multiple_species and
annotation files stored in an annotation folder for the query (one annotation file) and subject species in gtf or gff
file format, this function selects the best DIAMOND hit to represent either a gene locus (e.g. the splice variant of the gene locus with lowest e-value) or the best DIAMOND hit for a splice varaint." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">homologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import <code><a href='dnds.html'>dnds</a></code> tables by gene locus and splice varaint for a set of species</h1>
    
    <div class="hidden name"><code>import_dnds_across_multiple_species.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given an input folder storing dnds tables generated by <code><a href='dnds_across_multiple_species.html'>dnds_across_multiple_species</a></code> and
annotation files stored in an annotation folder for the query (one annotation file) and subject species in <code>gtf</code> or <code>gff</code>
file format, this function selects the best DIAMOND hit to represent either a gene locus (e.g. the splice variant of the gene locus with lowest e-value) or the best DIAMOND hit for a splice varaint.</p>
    </div>

    <pre class="usage"><span class='fu'>import_dnds_across_multiple_species</span>(
  <span class='kw'>dnds_output_folder</span>,
  <span class='kw'>annotation_file_query</span>,
  <span class='kw'>annotation_folder_subject</span>,
  <span class='kw'>output_folder</span>,
  output_type = <span class='st'>"gene_locus"</span>,
  format = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gtf"</span>, <span class='st'>"gtf"</span>)
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>dnds_output_folder</th>
      <td><p>file path to folder storing <code>dnds tables</code> generated with <code><a href='dnds_across_multiple_species.html'>dnds_across_multiple_species</a></code> and
stored conform with <code><a href='import_dnds_tbl.html'>import_dnds_tbl</a></code>.</p></td>
    </tr>
    <tr>
      <th>annotation_file_query</th>
      <td><p>file path to the annotation file of the query species in <code>gtf</code> or <code>gff</code> file format.</p></td>
    </tr>
    <tr>
      <th>annotation_folder_subject</th>
      <td><p>file path to a folder storing the annotation files of the subject species in <code>gtf</code> or <code>gff</code> file format.</p></td>
    </tr>
    <tr>
      <th>output_folder</th>
      <td><p>file path to a folder in which orthologs tables shall be stored.</p></td>
    </tr>
    <tr>
      <th>output_type</th>
      <td><p>type of ortholog table that shall be printed out (or stored in a variable).
Available options are:</p><ul>
<li><p><code>output_type = "gene_locus"</code> (default): find for each gene locus a representative splice variant that
maximizes the sequence homology (in terms of smallest e-value and longest splice variant in case of same evalue)
to the subject gene locus and its representative splice variant. The output table contains only once representative splice variant per gene locus.</p></li>
<li><p><code>output_type = "splice_variant"</code>: for each homologous gene locus determine for each splice variant their
respective splice variant homolog. he output table contains several splice variants and their homologous splice variants per gene locus.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>a vector of length 2 storing the annotation file formats of the query annotation file and subject annotation file: either <code>gtf</code> or <code>gff</code> format. E.g. <code>format = c("gtf","gtf")</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='dnds_across_multiple_species.html'>dnds_across_multiple_species</a></code>, <code><a href='dnds.html'>dnds</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hajk-Georg Drost</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>if</span> (<span class='fl'>FALSE</span>) {
<span class='co'># running dnds across several species using DIAMOND executable</span>
<span class='co'># path '/opt/miniconda3/bin/'</span>
<span class='fu'><a href='dnds_across_multiple_species.html'>dnds_across_multiple_species</a></span>(
  query      = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, package = <span class='st'>'homologr'</span>),
  subjects_folder = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/map_gen_example'</span>, package = <span class='st'>'homologr'</span>),
  diamond_exec_path = <span class='st'>"/opt/miniconda3/bin/"</span>,
  aa_aln_type      = <span class='st'>"pairwise"</span>,
  aa_aln_tool      = <span class='st'>"NW"</span>,
  codon_aln_tool   = <span class='st'>"pal2nal"</span>,
  dnds_estimation  = <span class='st'>"Li"</span>,
  output_folder   = <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(), <span class='st'>"homologr_dnds_maps"</span>),
  quiet           = <span class='fl'>TRUE</span>,
  cores           = <span class='fl'>1</span>
)
<span class='co'># Import dnds tables by gene locus and splice varaint for a set of species</span>
<span class='kw'>import_dnds_tables</span> <span class='op'>&lt;-</span> <span class='fu'>import_dnds_across_multiple_species</span>(
dnds_output_folder = <span class='st'>"homologr_dnds_maps"</span>,
annotation_file_query = <span class='st'>"system.file('seqs/ortho_thal.gtf', package = 'homologr')"</span>,
annotation_folder_subject = <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>'seqs/plants_subject_files'</span>, package = <span class='st'>'homologr'</span>),
output_folder = <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(), <span class='st'>"homologr_dnds_maps"</span>, <span class='st'>"orthologs_tables"</span>),
output_type = <span class='st'>"gene_locus"</span>,
format = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"gtf"</span>, <span class='st'>"gtf"</span>))

<span class='co'># look at results</span>
<span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>lapply</a></span>(<span class='kw'>import_dnds_tables</span>, <span class='kw'>head</span>)
}
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


