<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Internal dnds computations — internal_dnds • homologr</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Internal dnds computations — internal_dnds" />
<meta property="og:description" content="This function takes a vector containing the
query amino acid sequence, subject amino acid sequence, query CDS sequence, and subject CDS sequence
and then runs the following pipieline:

1) Multiple-Alignment of query amino acid sequence and subject amino acid sequence
2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence
3) dNdS estimation of the codon alignment returned by 2)

" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">homologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Internal dnds computations</h1>
    
    <div class="hidden name"><code>internal_dnds.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function takes a vector containing the
query amino acid sequence, subject amino acid sequence, query CDS sequence, and subject CDS sequence
and then runs the following pipieline:</p>
<ul>
<li><p>1) Multiple-Alignment of query amino acid sequence and subject amino acid sequence</p></li>
<li><p>2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence</p></li>
<li><p>3) dNdS estimation of the codon alignment returned by 2)</p></li>
</ul>

    </div>

    <pre class="usage"><span class='fu'>internal_dnds</span>(
  <span class='kw'>complete_tbl</span>,
  <span class='kw'>aa_aln_type</span>,
  <span class='kw'>aa_aln_tool</span>,
  <span class='kw'>codon_aln_tool</span>,
  <span class='kw'>dnds_estimation</span>,
  <span class='kw'>store_locally</span>,
  <span class='kw'>quiet</span>,
  <span class='kw'>cores</span>,
  <span class='kw'>clean_folders</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>complete_tbl</th>
      <td><p>a data.table object storing the query_id, subject_id, query_cds (sequence),
subject_cds (sequence), query_aa (sequence), and subject_aa (sequence) of the organisms that shall be compared.</p></td>
    </tr>
    <tr>
      <th>aa_aln_type</th>
      <td><p>a character string specifying the amino acid alignment type:  <code>aa_aln_type</code> = <code>"pairwise"</code>.
Default is <code>aa_aln_type</code> = "multiple".</p></td>
    </tr>
    <tr>
      <th>aa_aln_tool</th>
      <td><p>a character string specifying the multiple alignment tool that shall be used for pairwise protein alignments.</p></td>
    </tr>
    <tr>
      <th>codon_aln_tool</th>
      <td><p>a character string specifying the codon alignment tool that shall be used for codon alignments. Default is <code>codon_aln_tool</code> = <code>"pal2nal"</code>.</p></td>
    </tr>
    <tr>
      <th>dnds_estimation</th>
      <td><p>a character string specifying the dNdS estimation method, e.g. "Comeron". See Details for all options.</p></td>
    </tr>
    <tr>
      <th>store_locally</th>
      <td><p>a logical value indicating whether or not alignment files shall be stored locally rather than in <code><a href='https://rdrr.io/r/base/tempfile.html'>tempdir()</a></code>.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>a logical value specifying whether a successful interface call shall be printed out.</p></td>
    </tr>
    <tr>
      <th>cores</th>
      <td><p>a numeric value specifying the number of cores that shall be used to perform parallel computations on a multicore machine.</p></td>
    </tr>
    <tr>
      <th>clean_folders</th>
      <td><p>a boolean value spefiying whether all internall folders storing the output of used programs
shall be removed. Default is <code>clean_folders</code> = <code>FALSE</code>.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function takes the amino acid and CDS sequences two orthologous genes
and writes the corresponding amino acid and CDS sequences as fasta file into
the internal folder environment. The resulting fasta files (two files) store the
amino acid sequence of the query_id and subject_id (file one) and the CDS sequence of
the query_id and subject_id (file two). These fasta files are then used to pass through the following pipeline:</p>
<p>1) Multiple-Alignment or Pairwise-Alignment of query amino acid sequence and subject amino acid sequence</p>
<p>2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence</p>
<p>3) dNdS estimation of the codon alignment returned by 2)</p>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p><a href='http://www.r-bloggers.com/the-wonders-of-foreach/'>http://www.r-bloggers.com/the-wonders-of-foreach/</a></p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='substitutionrate.html'>substitutionrate</a></code>, <code><a href='dnds.html'>dnds</a></code></p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Hajk-Georg Drost</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


